File: packages/shared-types/src/insights.ts
import type { RequirementImportance, RoleFamily } from "./jobs";

/**
 * How much of a gap a skill is, from `docs/07`.
 *
 * Four states rather than a boolean, because "on the profile" is not the same
 * as "evidenced": a skill an import inferred is a claim, a skill the user
 * confirmed but attached to nothing is a claim they believe, and a skill
 * attached to real work is a record.
 */
export type GapState = "STRONG_GAP" | "PARTIAL_GAP" | "WEAK_EVIDENCE" | "NO_GAP";

/** What each state means, in the user's terms rather than the enum's. */
export const GAP_STATE_LABELS: Record<GapState, string> = {
  STRONG_GAP: "Not on your profile",
  PARTIAL_GAP: "Not confirmed by you",
  WEAK_EVIDENCE: "Listed, not demonstrated",
  NO_GAP: "Covered",
};

/** The sentence under each, saying what would close it. */
export const GAP_STATE_HINTS: Record<GapState, string> = {
  STRONG_GAP: "Nothing in your profile mentions it.",
  PARTIAL_GAP: "It came from an import and you have not confirmed it.",
  WEAK_EVIDENCE: "It is on your profile but not attached to any role or project.",
  NO_GAP: "Confirmed, and demonstrated in your work.",
};

export interface SkillDemandEntry {
  /** Canonical id where the catalogue knows the skill, normalised name where
      it does not. Stable enough to key a list on. */
  key: string;
  skill_id: string | null;
  name: string;
  /** Whether this resolved to the shared skill catalogue. */
  catalogued: boolean;
  /** How many of the user's analysed jobs asked for it. */
  jobs: number;
  /** That, as a percentage of the same set. */
  share: number;
  importance: Partial<Record<RequirementImportance, number>>;
  role_families: Record<string, number>;
  state: GapState;
  held: boolean;
}

/** Everything on this screen carries the sample it was counted over. */
interface Sampled {
  analysed_jobs: number;
  minimum_jobs: number;
  above_threshold: boolean;
}

export interface SkillDemandReport extends Sampled {
  skills: SkillDemandEntry[];
}

export interface SkillGapReport extends Sampled {
  gaps: SkillDemandEntry[];
}

export interface InsightsOverview extends Sampled {
  skills_tracked: number;
  gaps_found: number;
  top_skills: SkillDemandEntry[];
  top_gaps: SkillDemandEntry[];
}

/** One role family, and how the user's jobs in it have gone. */
export interface RoleInsight {
  role_family: RoleFamily;
  jobs: number;
  matched: number;
  /** Null below the threshold, and null when nothing in the family scored.
      Never zero — that would be a claim about fit, not about data. */
  average_alignment: number | null;
  applications: number;
}

export interface RoleReport {
  minimum_jobs: number;
  roles: RoleInsight[];
}

export interface FunnelStage {
  key: string;
  label: string;
  /** Applications that got *at least* this far, read from their history —
      not their current status, which would report every rejection as never
      having applied. */
  reached: number;
}

export interface FunnelReport {
  applications: number;
  minimum_applications: number;
  /** Whether the client may divide these counts. Decided server-side so the
      threshold lives in one place. */
  rates_are_meaningful: boolean;
  stages: FunnelStage[];
}

export interface ResumeInsight {
  resume_version_id: string;
  label: string;
  sent: number;
  reached_interview: number;
  offers: number;
}

export interface ResumePerformanceReport {
  minimum_applications: number;
  rates_are_meaningful: boolean;
  versions: ResumeInsight[];
}

File: apps/api/src/jip_api/api/v1/insights.py
"""`GET /api/v1/insights/*`.

All five routes ``docs/10-api-contracts.md`` lists, plus resume performance.

Role analysis shipped later than the rest and for a reason worth keeping: it was
held back on DEV-026, which said requirement importance moves between readings
and therefore poisons any average of alignment scores. Two five-reading
measurements found otherwise — keyed by resolved skill, importance moved once in
twenty-four labels across two postings, and what actually varies is which
requirements get extracted and how they are worded. That varies between
alternative readings of one posting, and in production a posting is read once.

So the blocker was never importance. It is sample size, which is the rule
``docs/07`` already states twice and which every figure here now carries:

- a share is quoted with the number of postings it was counted over;
- a role family's average alignment is null below three jobs;
- no conversion rate is computed below five applications, though the stage
  counts are still shown, because those are facts.

Nothing is ever zero where it means "not measured". Zero is a claim.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends
from pydantic import BaseModel, ConfigDict
from sqlalchemy.orm import Session

from jip_api.api.dependencies import CurrentUser
from jip_api.application.insights.demand import (
    MINIMUM_JOBS,
    DemandReport,
    SkillDemand,
    build_demand,
    build_gaps,
)
from jip_api.application.insights.gaps import GapState
from jip_api.application.insights.performance import (
    MINIMUM_APPLICATIONS,
    MINIMUM_JOBS_PER_ROLE,
    build_funnel,
    build_resume_performance,
    build_roles,
)
from jip_api.core.responses import DataResponse
from jip_api.domain.jobs.analysis import RequirementImportance, RoleFamily
from jip_api.infrastructure.db.session import get_session

router = APIRouter(prefix="/insights", tags=["insights"])

SessionDep = Annotated[Session, Depends(get_session)]


# --- payloads -----------------------------------------------------------------


class SkillDemandPayload(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    skill_id: uuid.UUID | None
    name: str
    catalogued: bool
    jobs: int
    share: int
    importance: dict[RequirementImportance, int]
    role_families: dict[str, int]
    state: GapState
    held: bool


class DemandPayload(BaseModel):
    analysed_jobs: int
    """The denominator. Always sent, so the client cannot render a share
    without the sample it came from."""

    minimum_jobs: int
    """The threshold, sent rather than duplicated in TypeScript — the number
    and the sentence explaining it have to agree, and two copies drift."""

    above_threshold: bool
    skills: list[SkillDemandPayload]


class GapsPayload(BaseModel):
    analysed_jobs: int
    minimum_jobs: int
    above_threshold: bool
    gaps: list[SkillDemandPayload]


class OverviewPayload(BaseModel):
    analysed_jobs: int
    minimum_jobs: int
    above_threshold: bool
    skills_tracked: int
    gaps_found: int
    top_skills: list[SkillDemandPayload]
    top_gaps: list[SkillDemandPayload]


# --- routes -------------------------------------------------------------------


@router.get(
    "/overview",
    response_model=DataResponse[OverviewPayload],
    summary="What the saved jobs add up to",
)
def read_overview(user: CurrentUser, session: SessionDep) -> DataResponse[OverviewPayload]:
    report = build_demand(session, user.id)
    gaps = build_gaps(report)

    return DataResponse(
        data=OverviewPayload(
            analysed_jobs=report.analysed_jobs,
            minimum_jobs=MINIMUM_JOBS,
            above_threshold=report.is_above_threshold,
            skills_tracked=len(report.skills),
            gaps_found=len(gaps),
            top_skills=_payloads(report.skills[:5]),
            top_gaps=_payloads(gaps[:5]),
        )
    )


@router.get(
    "/skills/demand",
    response_model=DataResponse[DemandPayload],
    summary="Which skills the saved jobs ask for",
)
def read_demand(user: CurrentUser, session: SessionDep) -> DataResponse[DemandPayload]:
    report = build_demand(session, user.id)

    return DataResponse(
        data=DemandPayload(
            analysed_jobs=report.analysed_jobs,
            minimum_jobs=MINIMUM_JOBS,
            above_threshold=report.is_above_threshold,
            skills=_payloads(report.skills),
        )
    )


@router.get(
    "/skills/gaps",
    response_model=DataResponse[GapsPayload],
    summary="Which of those the profile does not answer",
)
def read_gaps(user: CurrentUser, session: SessionDep) -> DataResponse[GapsPayload]:
    report: DemandReport = build_demand(session, user.id)

    return DataResponse(
        data=GapsPayload(
            analysed_jobs=report.analysed_jobs,
            minimum_jobs=MINIMUM_JOBS,
            above_threshold=report.is_above_threshold,
            gaps=_payloads(build_gaps(report)),
        )
    )


class RoleInsightPayload(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    role_family: RoleFamily
    jobs: int
    matched: int
    average_alignment: int | None
    """Null below the threshold, and null when nothing in the family scored.
    Never zero, which would be a claim about fit rather than about data."""

    applications: int


class RolesPayload(BaseModel):
    minimum_jobs: int
    roles: list[RoleInsightPayload]


class FunnelStagePayload(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    label: str
    reached: int


class FunnelPayload(BaseModel):
    applications: int
    minimum_applications: int
    rates_are_meaningful: bool
    """Whether the client may divide these counts. Sent rather than left to the
    frontend to decide, so the threshold lives in one place."""

    stages: list[FunnelStagePayload]


class ResumeInsightPayload(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    resume_version_id: uuid.UUID
    label: str
    sent: int
    reached_interview: int
    offers: int


class ResumePerformancePayload(BaseModel):
    minimum_applications: int
    rates_are_meaningful: bool
    versions: list[ResumeInsightPayload]


@router.get(
    "/roles",
    response_model=DataResponse[RolesPayload],
    summary="How the saved jobs group by role family",
)
def read_roles(user: CurrentUser, session: SessionDep) -> DataResponse[RolesPayload]:
    report = build_roles(session, user.id)

    return DataResponse(
        data=RolesPayload(
            minimum_jobs=MINIMUM_JOBS_PER_ROLE,
            roles=[RoleInsightPayload.model_validate(role) for role in report.roles],
        )
    )


@router.get(
    "/applications/funnel",
    response_model=DataResponse[FunnelPayload],
    summary="How far applications have got",
)
def read_funnel(user: CurrentUser, session: SessionDep) -> DataResponse[FunnelPayload]:
    report = build_funnel(session, user.id)

    return DataResponse(
        data=FunnelPayload(
            applications=report.applications,
            minimum_applications=MINIMUM_APPLICATIONS,
            rates_are_meaningful=report.rates_are_meaningful,
            stages=[FunnelStagePayload.model_validate(stage) for stage in report.stages],
        )
    )


@router.get(
    "/resumes",
    response_model=DataResponse[ResumePerformancePayload],
    summary="What happened after each resume version was sent",
)
def read_resume_performance(
    user: CurrentUser, session: SessionDep
) -> DataResponse[ResumePerformancePayload]:
    versions = build_resume_performance(session, user.id)

    return DataResponse(
        data=ResumePerformancePayload(
            minimum_applications=MINIMUM_APPLICATIONS,
            rates_are_meaningful=sum(v.sent for v in versions) >= MINIMUM_APPLICATIONS,
            versions=[ResumeInsightPayload.model_validate(version) for version in versions],
        )
    )


def _payloads(skills: list[SkillDemand]) -> list[SkillDemandPayload]:
    return [SkillDemandPayload.model_validate(skill) for skill in skills]
